### PR TITLE
Always send the cached token when user is logged in

### PR DESCRIPTION
**Pull request by Wauplin**
TLDR; this works with this PR:

```py
# Using cached token from `huggingface-cli login`
file = hf_hub_download("CompVis/stable-diffusion-v1-4", "README.md")
```

---
(cc @julien-c @osanseviero @LysandreJik @apolinario @patrickvonplaten @Pierrci)
Fix https://github.com/huggingface/huggingface_hub/issues/926.

(related to [this discussion](https://huggingface.slack.com/archives/C03V11RNS7P/p1663331447986929) (internal link), [this discussion](https://huggingface.slack.com/archives/C02EMARJ65P/p1663771575035239) and [this discussion](https://docs.google.com/document/d/1ZE0pM8kry56Qtob5mIAYuGi5ZKpR8zTNnekEvPHtDgU/edit) as well (internal link))

The goal is to have a centralized way to handle tokens in `huggingface_hub` (and downstream libraries) that causes as low friction as possible for the users.

~The idea here is to keep the rule "if token is not explicitly set and is not required then do not pass it in header so that the user is cannot be tracked on public repos". This PR aims to solve that by making a HEAD call to the Hub (without auth) to check if a repo is public or private. In case of a private repo, the token is then automatically sent to the Hub in all subsequent calls (the "private" property of a repo is cached).~

## EDIT: strategy has been revisited

The problem that has been formulated is that if a user wants to download files from a gated/private repos, they currently need to pass `use_auth_token=True` even though the user is logged in. This PR solves that by always sending the cached token if the user is logged in. This is a big change compared to the previous rule that was "don't send the token if not explicitly needed".  Hope is to reduce as much friction as possible.

This PR and the `build_hf_headers` helper aims also to harmonize how tokens are handled and validated across the platform.

**Examples:**
```py
# Get auth header
headers = build_hf_headers(use_auth_token="hf_***") # explicit token
headers = build_hf_headers(use_auth_token=True) # explicitly use cached token
headers = build_hf_headers(use_auth_token=False) # explicitly don't use cached token
headers = build_hf_headers() # implicit use of the cached token

DISABLE_IMPLICIT_HF_TOKEN=True # to set as env variable
headers = build_hf_headers() # token is not sent
```

```py
# Cannot use api_org tokens on write calls
headers = build_hf_headers(use_auth_token="api_org_***")
headers = build_hf_headers(use_auth_token="api_org_***", is_write_action=True) # throws an error !
```

(note: this PR also gets rid of the `datasets` dependency in the tests but that kinda unrelated. I got an issue with it that I reported [in this PR](https://github.com/huggingface/datasets/issues/4990))

(edit: once that's reviewed and merged, implementing https://github.com/huggingface/huggingface_hub/issues/1018 should be straightforward)

### CODE REVIEW

**Comment by lhoestq** [comment]
Cool ! I guess `datasets` can do the same at one point for consistency

**Comment by Wauplin** [comment]
@lhoestq Yes ! Best would be to handle all http calls in `huggingface_hub` but if `datasets` has a very specific call, it could reuse "build_hf_headers" (`from huggingface_hub.utils import build_hf_headers`).

File: src/huggingface_hub/utils/_hf_folder.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contain helper class to retrieve/store token from/to local cache."""
+import os
+from pathlib import Path
+from typing import Optional
+
+
+class HfFolder:
```
Comment (osanseviero) [inline src/huggingface_hub/utils/_hf_folder.py:21]: Are we ok with moving the location of `HfFolder`? Any backward compatibility issues we need to keep an eye at?

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from typing import Generator
+
+import pytest
+
+from huggingface_hub import HfFolder
+
+
+@pytest.fixture(autouse=True, scope="session")
```
Comment (osanseviero) [inline tests/conftest.py:8]: Nice!

**Review by osanseviero — APPROVED** [review]
Very clean PR, thanks a lot for doing this!

File: docs/source/quick-start.mdx [hunk 3]
```diff
@@ -82,9 +77,25 @@ Or if you prefer to work from a Jupyter or Colaboratory notebook, then login wit
...
+You can also provide your token to our functions and methods. This way you don't need to
+store your token anywhere.
+
+</Tip>
+
+<Tip>
+
+Once you are logged in, all requests to the Hub will use your access token by default.
+If you want to disable implicit use of your token, you should set the
+`DISABLE_IMPLICIT_HF_TOKEN` environment variable.
```
Comment (julien-c) [inline docs/source/quick-start.mdx:91]: i'm wondering if we don't have a convention of prefixing our env vars with `HF_`? 

what do the other env vars look like?

Comment (LysandreJik) [reply docs/source/quick-start.mdx:91]: From a quick scan we have:
- `HUGGINGFACE_CO_STAGING`
- `HF_ENDPOINT`
- `HF_HOME`
- `HUGGINGFACE_HUB_CACHE`
- `HF_HUB_OFFLINE`
- `HF_HUB_DISABLE_PROGRESS_BARS`

There's a striking lack of consensus, but I think preceding it by `HF_` would make sense: `HF_HUB_DISABLE_IMPLICIT_TOKEN` ?

File: src/huggingface_hub/utils/_headers.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        > build_hf_headers() # implicit use of the cached token
+        {"authorization": "Bearer hf_***"}
+
+        # DISABLE_IMPLICIT_HF_TOKEN=True # to set as env variable
+        > build_hf_headers() # token is not sent
+        {}
+
+        > build_hf_headers(use_auth_token="api_org_***", is_write_action=True)
+        ValueError: You must use your personal account token for write-access methods.
```
Comment (LysandreJik) [inline src/huggingface_hub/utils/_headers.py:70]:
```suggestion
        >>> build_hf_headers(use_auth_token="hf_***") # explicit token
        {"authorization": "Bearer hf_***"}

        >>> build_hf_headers(use_auth_token=True) # explicitly use cached token
        {"authorization": "Bearer hf_***"}

        >>> build_hf_headers(use_auth_token=False) # explicitly don't use cached token
        {}

        >>> build_hf_headers() # implicit use of the cached token
        {"authorization": "Bearer hf_***"}

        # DISABLE_IMPLICIT_HF_TOKEN=True # to set as env variable
        >>> build_hf_headers() # token is not sent
        {}

        >>> build_hf_headers(use_auth_token="api_org_***", is_write_action=True)
        ValueError: You must use your personal account token for write-access methods.
```

File: src/huggingface_hub/utils/_headers.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    # Get auth token to send
+    token_to_send = _get_token_to_send(use_auth_token)
+    _validate_token_to_send(token_to_send, is_write_action=is_write_action)
+
+    # Combine headers
+    headers = {}
+    if token_to_send is not None:
+        headers["authorization"] = f"Bearer {token_to_send}"
+    # TODO: add user agent in headers
```
Comment (LysandreJik) [inline src/huggingface_hub/utils/_headers.py:89]: :+1:

File: src/huggingface_hub/utils/_hf_folder.py (hunk 1 above)
Comment (LysandreJik) [reply src/huggingface_hub/utils/_hf_folder.py:21]: We could reimport it in the `hf_api.py` file just to make sure BC is respected. I agree it's important given that in this libary we have private modules (`_deprecation.py`, `_headers.py`, ...) which sends the signal that public modules should have stable import APIs

File: tests/conftest.py (hunk 2 above)
Comment (LysandreJik) [reply tests/conftest.py:8]: That's extremely clean

**Review by LysandreJik — APPROVED** [review]
Cool! The resulting use of `build_hf_headers` is nice, much cleaner this way.
Thanks for working on it!

File: docs/source/quick-start.mdx (hunk 3 above)
Comment (Wauplin) [reply docs/source/quick-start.mdx:91]: `HF_HUB_DISABLE_IMPLICIT_TOKEN` sounds definitely better ! Thanks for raising this and proposing a solution :)

File: src/huggingface_hub/utils/_hf_folder.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/utils/_hf_folder.py:21]: Nice catch @osanseviero , thanks for reporting.

Actually it would have break [here](https://github.com/huggingface/transformers/blob/c7fd28999fb8a03f4839ab5402a7c32997ded0fa/src/transformers/commands/user.py#L19) in `transformers`: . I made the re-import in `hf_api.py` as suggested by @LysandreJik.

For the record, I needed to remove `HfFolder` from `hf_api.py` to avoid circular imports between `utils._headers.py` and `hf_api.py`.

File: docs/source/quick-start.mdx [hunk 6]
```diff
@@ -54,21 +54,16 @@ full-length hash instead of the shorter 7-character commit hash:
...
-your Settings. The User Access Token is used to authenticate your identity to the Hub.
+## Login
 
-<Tip>
+In a lot of cases, you must be logged in with a Hugging Face account to interact with
+the Hub: download private repos, upload files, create PRs,...
+[Create an account](https://hf.co/join) if you don't already have one, and then sign in
+to get your [User Access Token](https://huggingface.co/docs/hub/security-tokens) from
+your [Settings page](https://huggingface.co/settings/tokens). The User Access Token is
+used to authenticate your identity to the Hub.
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/quick-start.mdx:64]: Good idea, it doesn't cost anything to add that. I've added a `<Tip>` in the "Create a repo" section.

File: src/huggingface_hub/utils/_headers.py (hunk 4 above)
Comment (Wauplin) [reply src/huggingface_hub/utils/_headers.py]: :+1:

**Comment by Wauplin** [comment]
Thanks @julien-c @LysandreJik @osanseviero for the reviews and comments. I made all the requested changes.
Will merge it once the CI is happy :)

File: src/huggingface_hub/utils/_headers.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            raise EnvironmentError(
+                "Token is required (`use_auth_token=True`), but no token found. You"
+                " need to provide a token or be logged in to Hugging Face with"
+                " `huggingface-cli login` or `notebook_login`. See"
+                " https://huggingface.co/settings/tokens."
+            )
+        return cached_token
+
+    # Case implicit use of the token is forbidden by env variable
+    if os.environ.get("HF_HUB_DISABLE_IMPLICIT_TOKEN"):
```
Comment (lhoestq) [inline src/huggingface_hub/utils/_headers.py:118]: (nit) In `datasets` and `transformers`, we usually check if the value is not in
```python
ENV_VARS_TRUE_VALUES = {"1", "ON", "YES", "TRUE"}
```

Comment (Wauplin) [reply src/huggingface_hub/utils/_headers.py:118]: Addressed in https://github.com/huggingface/huggingface_hub/pull/1081
